src/Dofs/sparsity_pattern.jl

```diff
@@ -102,38 +401,164 @@ function _coupling_to_local_dof_coupling(dh::DofHandler, coupling::AbstractMatri
+                        end
+                        r = insert_sorted(r, col)
+                        sp′[row′] = r
+                    end
+                else
+                    # ... this column _is_ constrained, double-distribute to columns/rows.
+                    for (row′, _) in row_coeffs
+                        !keep_constrained && haskey(dofmapping, row′) && continue
+                        for (col′, _) in col_coeffs
+                            !keep_constrained && haskey(dofmapping, col′) && continue
+                            r = get(sp′, row′) do
+                                HeapAllocator.resize(HeapAllocator.alloc_array(heap, Int, 8), 0)
+                            end
+                            r = insert_sorted(r, col′)
+                            sp′[row′] = r
+                        end
+                    end
+                end
+            end
+        end
+    end
+
+    # Insert new entries into the sparsity pattern
+    for (row, colidxs) in sp′
+        # TODO: Extract row here and just insert_sorted
+        for col in colidxs
+            add_entry!(sp, row, col)
+        end
+    end
+
+    # Release the memory in the temporary heap
+    HeapAllocator.free(heap)
+
+    return sp
+end
+
+function _add_cross_coupling(sp::SparsityPattern, coupling_sdh::Matrix{Bool}, dof_i::Int, dof_j::Int,
         cell_field_dofs::Union{Vector{Int}, SubArray}, neighbor_field_dofs::Union{Vector{Int}, SubArray},
-        i::Int, j::Int, sym::Bool, keep_constrained::Bool, ch::Union{ConstraintHandler, Nothing}, cnt::Int, I::Vector{Int}, J::Vector{Int})
+        i::Int, j::Int, keep_constrained::Bool, ch::Union{ConstraintHandler, Nothing})
 
-    coupling_sdh[dof_i, dof_j] || return cnt
+    coupling_sdh[dof_i, dof_j] || return
     dofi = cell_field_dofs[i]
     dofj = neighbor_field_dofs[j]
-    sym && (dofj > dofi && return cnt)
-    !keep_constrained && (haskey(ch.dofmapping, dofi) || haskey(ch.dofmapping, dofj)) && return cnt
-    cnt += 1
-    _add_or_grow(cnt, I, J, dofi, dofj)
-    return cnt
+    # sym && (dofj > dofi && return cnt)
+    !keep_constrained && (haskey(ch.dofmapping, dofi) || haskey(ch.dofmapping, dofj)) && return
+    add_entry!(sp, dofi, dofj)
+    return
 end
 
 """
-    cross_element_coupling!(dh::DofHandler, topology::ExclusiveTopology, sym::Bool, keep_constrained::Bool, couplings::Union{AbstractVector{<:AbstractMatrix{Bool}},Nothing}, cnt::Int, I::Vector{Int}, J::Vector{Int})
-
-Mutates `I, J` to account for cross-element coupling by calling [`_add_cross_coupling`](@ref).
-Returns the updated value of `cnt`.
+    cross_element_coupling!(
+        sp::SparsityPattern, dh::DofHandler, ch::Union{ConstraintHandler, Nothing},
+        topology::ExclusiveTopology, keep_constrained::Bool,
+        cross_coupling::AbstractMatrix{Bool},
+    )
 
-Used internally for sparsity patterns with cross-element coupling.
+TODO: Expose to users?
```

Comment:
I would lean towards answering this with a "yes" to give users more fine grained control (and modularity) over the sparsity pattern. And maybe we should discuss the name again in this case.